File: week1_basic_data_structures/3_network_simulation/process_packages.java
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Scanner;

class Request {
    public int arrival_time;
    public int process_time;

    public Request(int arrival_time, int process_time) {
        this.arrival_time = arrival_time;
        this.process_time = process_time;
    }
}

class Response {

    public boolean dropped;
    public int start_time;
    public int finish_time;

    public Response(boolean dropped, int start_time, int finish_time) {
        this.dropped = dropped;
        this.start_time = start_time;
        this.finish_time = finish_time;
    }
}

class Buffer {
    private int size;
    private ArrayDeque<Response> q;

    public Buffer(int size) {
        this.size = size;
        this.q = new ArrayDeque<Response>(size);
    }

    public Response process(Request req) {
        // write your code here
        // return process_packages.processNext(size, request, q);
        int arrival_time = req.arrival_time;
        while (!q.isEmpty() && q.peekFirst().finish_time <= arrival_time)
            q.remove();
        if (q.size() == size)
            return new Response(true, -1, 0);
        int next_available_time = q.isEmpty() ? -1 : q.peekLast().finish_time;
        int start_time = Math.max(arrival_time, next_available_time);
        int finish_time = start_time + req.process_time;
        Response r = new Response(false, start_time, finish_time);
        q.add(r);
        return r;
    }
}

class process_packages {

    // Mutates requestQueue to add new arrival, if within capacity
    static Response processNext(int size, Request req, ArrayDeque<Response> q) {
        int arrival_time = req.arrival_time;
        while (!q.isEmpty() && q.peekFirst().finish_time <= arrival_time)
            q.remove();
        if (q.size() == size)
            return new Response(true, -1, 0);
        int next_available_time = q.isEmpty() ? -1 : q.peekLast().finish_time;
        int start_time = Math.max(arrival_time, next_available_time);
        int finish_time = start_time + req.process_time;
        Response r = new Response(false, start_time, finish_time);
        q.add(r);
        return r;
    }

    private static ArrayList<Request> ReadQueries(Scanner scanner) {
        int requests_count = scanner.nextInt();
        ArrayList<Request> requests = new ArrayList<Request>();
        for (int i = 0; i < requests_count; ++i) {
            int arrival_time = scanner.nextInt();
            int process_time = scanner.nextInt();
            requests.add(new Request(arrival_time, process_time));
        }
        return requests;
    }

    private static ArrayList<Response> ProcessRequests(ArrayList<Request> requests, Buffer buffer) {
        ArrayList<Response> responses = new ArrayList<Response>();
        for (int i = 0; i < requests.size(); ++i) {
            responses.add(buffer.process(requests.get(i)));
        }
        return responses;
    }

    private static void PrintResponses(ArrayList<Response> responses) {
        for (int i = 0; i < responses.size(); ++i) {
            Response response = responses.get(i);
            if (response.dropped) {
                System.out.println(-1);
            } else {
                System.out.println(response.start_time);
            }
        }
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int buffer_max_size = scanner.nextInt();
        Buffer buffer = new Buffer(buffer_max_size);
        ArrayList<Request> requests = ReadQueries(scanner);
        ArrayList<Response> responses = ProcessRequests(requests, buffer);
        PrintResponses(responses);
    }
}

File: week1_basic_data_structures/3_network_simulation/process_packages.py
# python3

from collections import namedtuple, deque

Request = namedtuple("Request", ["arrived_at", "process_time"])
Response = namedtuple("Response", ["dropped", "started_at", "finished_at"])


class BoundedDeque:
    def __init__(self, size):
        self.size = size
        self.len = 0
        self.q = deque()

    def empty(self):
        return self.len == 0

    def full(self):
        return self.len == self.size

    def add(self, elem):
        if self.len == self.size:
            raise IndexError("Added to full BoundedDeq")
        self.len += 1
        self.q.appendleft(elem)

    def peek_last(self):
        if self.len == 0:
            raise IndexError("Peeked empty BoundedDeq")
        r = self.q.popleft()
        self.q.appendleft(r)
        return r

    def take(self):
        if self.len == 0:
            raise IndexError("Took empty BoundedDeq")
        self.len -= 1
        return self.q.pop()

    def drop_until(self, pred) -> None:
        # leaves first passing element in queue
        while True:
            try:
                nxt = self.take()
            except IndexError:
                break
            if pred(nxt):
                # put it back
                self.q.append(nxt)
                self.len += 1
                break


class Buffer:
    def __init__(self, size):
        self.size = size
        self.q = BoundedDeque(size)
        self.finish_time = []

    def process(self, req: Request) -> Response:
        arrival_time = req.arrived_at
        self.q.drop_until(lambda resp: resp.finished_at > arrival_time)
        r = None
        if self.q.full():
            # dropped, not added
            r = Response(True, None, None)
        elif self.q.empty():
            r = Response(False, arrival_time, arrival_time + req.process_time)
            self.q.add(r)
        else:
            last = self.q.peek_last()
            r = Response(False,
                         last.finished_at,
                         last.finished_at + req.process_time)
            self.q.add(r)
        return r


def process_requests(requests, buffer):
    responses = []
    for request in requests:
        responses.append(buffer.process(request))
    return responses


def main():
    buffer_size, n_requests = map(int, input().split())
    requests = []
    for _ in range(n_requests):
        arrived_at, time_to_process = map(int, input().split())
        requests.append(Request(arrived_at, time_to_process))
    buffer = Buffer(buffer_size)
    responses = process_requests(requests, buffer)
    for response in responses:
        print(response.started_at if not response.dropped else -1)


if __name__ == "__main__":
    main()
